### ISSUE TITLE
Firestore crashes after `10.7.1` -> `10.7.2` update in Expo/React Native

### ISSUE THREAD

**Issue by Yonom** [firebase/firebase-js-sdk#7962]
### Operating System

iOS 17.2

### Browser Version

React Native 0.73.2 / Hermes

### Firebase SDK Version

10.7.2

### Firebase SDK Product:

Firestore

### Describe your project's tooling

React Native app initialized with Expo

### Describe the problem

After upgrading to Firestore SDK 10.7.2, fetching documents from Firestore causes an error.

The crash occurs during the `isSafari()` call introduced in this commit: https://github.com/firebase/firebase-js-sdk/commit/d7ace80d44ec870c3117cfed04ae6a1988c03c8e#diff-d73aa623e0442e31903665f2d555bbf0e1a317db2009f526e8f7209ebf662917R66 

This is because the `isSafari()` function from `@firestore/utils` accesses `navigator.userAgent`. On React Native, `navigator.userAgent` is `undefined`.

### Steps and code to reproduce issue

```
@firebase/firestore: Firestore (10.7.2): INTERNAL UNHANDLED ERROR:  TypeError: Cannot read property 'includes' of undefined
    at isSafari
    at __PRIVATE_getDefaultRelativeIndexReadCostPerDocument
    at __PRIVATE_QueryEngine
    at createLocalStore
```

**Comment by WhyGolfRyan** [comment]
+1

**Comment by LoranMaes** [comment]
I'm getting this error too and I've been stuck with it for the past day... I have to send this project to my teachers as my final and can't get it to work

**Comment by brunoqueiros** [comment]
@LoranMaes Remove `node_modules` and `package-lock.json` and rollback `firebase` to `10.7.1`

**Comment by Yonom** [comment]
Another temporary hacky fix would be to set userAgent :p

somewhere before calling firebase functions, e.g. in `index.js`:
```ts
window.navigator.userAgent = "ReactNative"; // javascript
// or
(window.navigator as any).userAgent = "ReactNative"; // typescript
```

**Comment by themtaysw** [comment]
wtf... but my version:  "firebase": "^10.3.1", and i am getting this bug just randomly appeared out of nowhere now. what did you guy do?!

**Comment by Peege151** [comment]
I'm getting the same error with 10.7.1...

```
@firebase/firestore: Firestore (10.7.1): INTERNAL UNHANDLED ERROR:  TypeError: Cannot read property 'includes' of undefined
    at isSafari 
```

**Comment by dconeybe** [comment]
@Peege151 That's really odd that you're seeing the same crash on 10.7.1. That shouldn't be possible. Make sure you clear out your `node_modules` directory and any build caches that you have. Regardless, the fix to `isSafari()` will be included in the next release and should fix whatever problem you're seeing too.

**Comment by Yonom** [comment]
@Peege151 make sure you've specified "10.7.1" and not "^10.7.1" or "~10.7.1" in your package.json. To be extra sure, check your yarn.lock file to see if the package actually resolves to "10.7.1"

**Comment by kaberaarthur** [comment]
I have a different problem; when I roll back to 10.7.1, I am getting the error "Error: Service firestore is not available". The code was running perfectly on version 10.0.0 but now am getting the  "Error: Service firestore is not available" error with that one too. Now I am just really lost, here is my firebase package - "firebase": "10.7.1" - and here is my firebase config - 


import firebase from "firebase/compat/app";
import "firebase/compat/auth";
import "firebase/compat/firestore";

// Your web app's Firebase configuration
// For Firebase JS SDK v7.20.0 and later, measurementId is optional
const firebaseConfig = {
 ...
};

// Initialize Firebase
const app = firebase.initializeApp(firebaseConfig);

// Use these for db & auth
const db = app.firestore();
const auth = firebase.auth();

export { auth, db };

**Comment by dev-arekusandoru** [comment]
@kaberaarthur getting the same issue. 10.7.2 gives me the error `@firebase/firestore: Firestore (10.7.2): INTERNAL UNHANDLED ERROR:  TypeError: Cannot read property 'includes' of undefined`. 

reverting back to 10.7.1 gives me the error `Error: Service database is not available`

**Comment by lasithalakmal** [comment]
Getting the same error.

**Comment by Yonom** [comment]
The bug reported in this PR has been fixed (pending release). In the meantime, use my [workaround](https://github.com/firebase/firebase-js-sdk/issues/7962#issuecomment-1902290249) or move away from `10.7.2`.

@kaberaarthur @dev-arekusandoru  @lasithalakmal If you're facing other issues with other versions, open a new issue for that. It's off topic here...

**Comment by mamadouD29** [comment]
@dev-arekusandoru 
@lasithalakmal 
@Yonom 
@arsotech 
here is what I actually did to fix it when I rolled back.

1. I Stopped my expo project
2. In my package.json I removed the "firebase": "^10.7.2" line
3. I deleted the package-lock.json and node_modules
4. I install firebase 10.7.1 : npm install firebase@10.7.1 , it will install all your node_modules and package-lock.json
5. I started my project back  and all work pretty good like nothing happened
Give a try and let me know.

**Comment by tansiner** [comment]
> @dev-arekusandoru @lasithalakmal @Yonom @arsotech here is what I actually did to fix it when I rolled back.
> 
>     1. I Stopped my expo project
> 
>     2. In my package.json I removed the "firebase": "^10.7.2" line
> 
>     3. I deleted the package-lock.json and node_modules
> 
>     4. I install firebase 10.7.1 : npm install firebase@10.7.1 , it will install all your node_modules and package-lock.json
> 
>     5. I started my project back  and all work pretty good like nothing happened. npx expo start -c
> 
> 
> That is the solution right. Give a try

only removed package-lock.json, "firebase": "^10.7.2" line from package.json and then installed earlier version which is firebase@10.7.1, it worked like a charm, thanks a lot, you are the man

**Comment by arsotech** [comment]
Yes that worked thank you

**Comment by SevostyanovAleksandr** [comment]
@dev-arekusandoru @lasithalakmal @Yonom @arsotech, помогло решение

**Comment by kaberaarthur** [comment]
I finally solved my issue, but I did that by creating a new project and moving all my code into it. I bet this is a rookie solution😂, but the app is now running ok on Firebase 10.7.1

**Comment by Yonom** [comment]
@dconeybe counting everyone who interacted with this issue, this issue has affected 25+ devs and counting...

Would it be possible to accelerate a release cut? 😊

**Comment by Yonom** [comment]
Also, why are y'all doing convoluted things to downgrade your package?!

`npm i firebase@10.7.1`
or 
`yarn add firebase@10.7.1`

then restart expo server; that should be sufficient...

**Comment by KennethAduan** [comment]
> I have a different problem; when I roll back to 10.7.1, I am getting the error "Error: Service firestore is not available". The code was running perfectly on version 10.0.0 but now am getting the "Error: Service firestore is not available" error with that one too. Now I am just really lost, here is my firebase package - "firebase": "10.7.1" - and here is my firebase config -
> 
> import firebase from "firebase/compat/app"; import "firebase/compat/auth"; import "firebase/compat/firestore";
> 
> // Your web app's Firebase configuration // For Firebase JS SDK v7.20.0 and later, measurementId is optional const firebaseConfig = { ... };
> 
> // Initialize Firebase const app = firebase.initializeApp(firebaseConfig);
> 
> // Use these for db & auth const db = app.firestore(); const auth = firebase.auth();
> 
> export { auth, db };

same Issue with this

**Comment by Yonom** [comment]
A new firebase version was released, I guess you can now upgrade to that one for the fix: `npm i firebase@10.8.0`

**Comment by dconeybe** [comment]
Firebase JS SDK version 10.8.0 was released on February 1, 2024 with the fix for this bug (https://firebase.google.com/support/release-notes/js#version_1080_-_february_1_2024). Please upgrade to pick up the fix. Again, apologies for the inconvenience this bug caused.
